Assets/MixedRealityToolkit.Tests/PlayModeTests/PointerEventsTests.cs

```diff
@@ -0,0 +1,313 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#if !WINDOWS_UWP
+// When the .NET scripting backend is enabled and C# projects are built
+// Unity doesn't include the the required assemblies (i.e. the ones below).
+// Given that the .NET backend is deprecated by Unity at this point it's we have
+// to work around this on our end.
+using NUnit.Framework;
+using UnityEngine;
+using UnityEngine.TestTools;
+using System.Collections;
+using UnityEditor;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using Microsoft.MixedReality.Toolkit.Input;
+
+namespace Microsoft.MixedReality.Toolkit.Tests
+{
+    // Tests to verify that pointer events are being raised correctly
+    public class PointerEventsTests
+    {
+
+        // Helper script used to keep track of the number of pointer events raised
+        private class PointerHandler : MonoBehaviour, IMixedRealityFocusChangedHandler, IMixedRealityFocusHandler, IMixedRealityPointerHandler
+        {
+            public class State
+            {
+                public int numBeforeFocusChange;
+                public int numFocusChanged;
+                public int numFocusEnter;
+                public int numFocusExit;
+                public int numPointerDown;
+                public int numPointerUp;
+                public int numPointerClicked;
+                public int numPointerDragged;
+
+                public void AssertEqual(State expected)
+                {
+                    Assert.AreEqual(expected.numBeforeFocusChange, numBeforeFocusChange);
+                    Assert.AreEqual(expected.numFocusChanged, numFocusChanged);
+                    Assert.AreEqual(expected.numFocusEnter, numFocusEnter);
+                    Assert.AreEqual(expected.numFocusExit, numFocusExit);
+                    Assert.AreEqual(expected.numPointerDown, numPointerDown);
+                    Assert.AreEqual(expected.numPointerUp, numPointerUp);
+                    Assert.AreEqual(expected.numPointerClicked, numPointerClicked);
+                    Assert.AreEqual(expected.numPointerDragged, numPointerDragged);
+                }
+            }
+
+            public State state = new State();
+            private int numFocusing;
+
+            void IMixedRealityFocusChangedHandler.OnBeforeFocusChange(FocusEventData eventData)
+            {
+                Assert.AreNotSame(eventData.OldFocusedObject, eventData.NewFocusedObject);
+                Assert.AreEqual(state.numFocusChanged, state.numBeforeFocusChange);
+
```

Comment:
Can we assert numFocusEnter == numFocusExit here? Or should it be numFocusEnter  == numFocusExit + 1, i.e. already entered?

Similarly i would expect numFocusEnter == numFocusExit + 1 in OnFocusChanged.

Reply:
Focus enter/left events are received only for the first pointer entering/last leaving.

Reply:
I see, never mind then.